### ISSUE TITLE
Handle subcommands (without dashes) in OptionParser

### ISSUE THREAD

**Issue by bew** [crystal-lang/crystal#3357]
I'm trying to make a CLI app with subcommands like crystal init/build/run subcommands (without dashes), but not like crystal way ([crystal/command.cr#L65](https://github.com/crystal-lang/crystal/blob/master/src/compiler/crystal/command.cr#L65)), but using OptionParser only (with nested parsers)
So I need to use (at least) 2 nested OptionParser, and options without dashes and options.

``` crystal
require "option_parser"

args = ["one", "subcommand", "two", "three"]

OptionParser.parse args do |parser|
  parser.on "subcommand", "message" do

    # args == ["two", "three"]
    OptionParser.parse args do |parser|
      parser.on "two ARG", "description" do |arg|
        # never go here
      end

      parser.on "verbose", "description" do
        # go here
      end
    end
  end
end
```

I'm on Crystal v0.18.7 on Archlinux

It seems that `parser.on "option VALUE"` never match, but `parser.on "--option VALUE"` will..
It it intentionnal ?

**Comment by asterite** [comment]
Yes, OptionParser doesn't support commands and subcommands. The way to do it is to switch over the first element in ARGV, like what the compiler does. Alternatively there are some shards that provide this behaviour, for example [commander](https://github.com/mrrooijen/commander) (but I personally think switching over a string is really simple and works quite well).

**Comment by Bew78LesellB** [comment]
Thanks for the link, but it seems too compllicated and not as readable as OptionParser..
As stated here : [OptionParse#on][OP#on] : 
> Flags can (but don't have to) start with a dash. They can also have an optional argument, which will get passed to the block.

Flags without dashes should be able to have an optionnal argument, but as told earlier, `parser.on("flag OPTION", "description", &block)` doesn't work..
I don't think it's the wanted behavior, is it ?

[OP#on]: https://crystal-lang.org/api/0.18.7/OptionParser.html#on%28flag%2Cdescription%2C%26block%3AString-%3E%29-instance-method

**Comment by asterite** [comment]
I think it's pretty standard that command line options must start with a dash. Ruby's OptionParser also doesn't support what you are trying to do. We can probably raise an exception if the given string doesn't start with a dash...

**Comment by chaniks** [comment]
I actually love the current implementation.  It is fully customisable.

However, as he showed, the document might be misleading.
> Flags can (but don't have to) start with a dash. ***"They"*** can also have an optional argument, which will get passed to the block.

For me, it sounds more like; flags without dashes can also have optional argument—which is currently not true.

**Comment by straight-shoota** [comment]
Resolved by #9009
